### ISSUE TITLE
Gitpoller: command 'rev-parse' failed with 'unknown revision or path not in the working tree'

### ISSUE THREAD

**Issue by antonysigma** [buildbot/buildbot#6591]
Buildbot version: 3.4.0

I am getting this error when I was setting up the change_source to a private git repo:
```
buildbot.changes.gitpoller.GitError: command ['rev-parse', 'refs/buildbot/%2Fvar%2Fbuildbot%2Fabc/my_branch'] in /var/buildbot/master/gitpoller-work on repourl /var/buildbot/abc failed with exit code 128: fatal: ambiguous argument 'refs/buildbot/%2Fvar%2Fbuildbot%2Fabc/my_branch': unknown revision or path not in the working tree.
        Use '--' to separate paths from revisions, like this:
        'git <command> [<revision>...] -- [<file>...]
```

What is puzzling is that when I `cd` into the `/var/buildbot/master/gitpoller-work` directory and run the exact same command `git rev-parse refs/buildbot/%2Fvar%2Fbuildbot%2Fabc/my_branch` every ~3 seconds. Sometimes it prints the SHA on the stdout, sometimes return `unknown revision...`.

I am wondering if there is a race condition between `git fetch` command and `git rev-parse` command? How do I get the Git poller setup properly?

Here's the code to setup the Gitpoller on my `master.cfg`:
```python
c["change_source"].append(changes.GitPoller(
    repourl='/var/buildbot/abc',
    branches=['my_branch',],
    pollInterval = 10,
    pollAtLaunch = True,
    buildPushesWithNoCommits=True,
))
```

**Comment by SebKuzminsky** [comment]
Something similar is affecting me, using buildbot 3.6.1:
~~~
c['change_source'].append(
    changes.GitPoller(
        repourl='git://10.0.0.4/linuxcnc-upstream.git',
        branches=True,
        pollInterval=10
    )
)
~~~

I get this in the log every time the gitpoller runs:
~~~
2022-12-16 08:43:25-0700 [-] gitpoller: processing changes from "git://10.0.0.4/linuxcnc-upstream.git"
2022-12-16 08:43:25-0700 [-] trying to poll branch refs/heads/2.8 of git://10.0.0.4/linuxcnc-upstream.git
        Traceback (most recent call last):
          File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 857, in _runCallbacks
            current.result = callback(  # type: ignore[misc]
          File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1750, in gotResult
            current_context.run(_inlineCallbacks, r, gen, status)
          File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1656, in _inlineCallbacks
            result = current_context.run(
          File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 514, in throwExceptionIntoGenerator
            return g.throw(self.type, self.value, self.tb)
        --- <exception caught here> ---
          File "/usr/lib/python3/dist-packages/buildbot/changes/gitpoller.py", line 265, in poll
            rev = yield self._dovccmd(
          File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1656, in _inlineCallbacks
            result = current_context.run(
          File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 514, in throwExceptionIntoGenerator
            return g.throw(self.type, self.value, self.tb)
          File "/usr/lib/python3/dist-packages/buildbot/changes/gitpoller.py", line 449, in _dovccmd
            stdout = yield self._dovccmdImpl(command, args, path, None)
          File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1660, in _inlineCallbacks
            result = current_context.run(gen.send, result)
          File "/usr/lib/python3/dist-packages/buildbot/changes/gitpoller.py", line 479, in _dovccmdImpl
            raise GitError(f'command {full_args} in {path} on repourl {self.repourl} failed '
        buildbot.changes.gitpoller.GitError: command ['rev-parse', 'refs/buildbot/git%3A%2F%2F10.0.0.4%2Flinuxcnc-upstream.git/2.8'] in /var/lib/buildbot/masters/linuxcnc/gitpoller-work on repourl git://10.0.0.4/linuxcnc-upstream.git failed with exit code 128: fatal: ambiguous argument 'refs/buildbot/git%3A%2F%2F10.0.0.4%2Flinuxcnc-upstream.git/2.8': unknown revision or path not in the working tree.
        Use '--' to separate paths from revisions, like this:
        'git <command> [<revision>...] -- [<file>...]'
~~~

I get one copy of that traceback for each branch in the repo.

Cloning the repourl by hand works fine.  

Like @antonysigma reported, running `git rev-parse` on the ref by hand in the gitpoller-work dir works intermittently.

I tried stopping my buildbot, deleting the gitpoller-work dir, and restarting the buildbot.  The first poll works fine, the second poll and all after it fail in this way.

**Comment by shareefj** [comment]
I've just updated from an old version of Buildbot and hit this issue.  Is there any planned fix or should I just roll back?

**Comment by antonysigma** [comment]
> I am wondering if there is a race condition between git fetch command and git rev-parse command? How do I get the Git poller setup properly?

Sharing a mitigation to resolve the issue... I found that if I increased the `pollInterval` from 10 seconds to 10 minutes or more, the issue resolves on its own.

Perhaps the `GitPoller`, by design, assumes the HTTP/SSH response latency from the external git server is less than the value set in `pollInterval` argument.

**Comment by shareefj** [comment]
So after rolling back to my Buildbot revision I'm still seeing this error but only in the test instance I'm running.  Our production instance doesn't seem to have any of these rev-parse errors.  No idea what the difference is yet.

While investigating I managed to see a rev-parse error within the gitpoller-work directory which seemed to be caused by all refs being deleted and re-populated.  Does the GitPoller start from a clean work folder each time?  Or should it be bombing out early if the previous invocation is still running?

**Comment by shareefj** [comment]
So I've managed to replicate what's going on with a pure Git command.  After lots of `log.msg()` printing, I noticed that the refs in `refs/buildbot/<escaped_url>/*` were disappearing every other time the `git fetch` command was called.  I'm running Git 2.18.0.

So as an example, running the following command consecutively will fail every other try.

```
$ git fetch https://<bitbucket_url>.git +<branch>:refs/buildbot/<escaped_rul>/<branch>; git rev-parse refs/buildbot/<escaped_url>/<branch>
```

This doesn't happen with Git 1.8.3.1 which is the only other version I have on my system.  When it fails I get the following:

```
 - [deleted]           (none)     -> refs/buildbot/<escaped_url>/<branch>
```
